diff --git a/debugfs/dump.c b/debugfs/dump.c
index cc81a50..089e75d 100644
--- a/debugfs/dump.c
+++ b/debugfs/dump.c
@@ -52,6 +52,9 @@ static struct {
 	{ LINUX_S_IROTH, S_IROTH },
 	{ LINUX_S_IWOTH, S_IWOTH },
 	{ LINUX_S_IXOTH, S_IXOTH },
+	{ LINUX_S_ISUID, S_ISUID },
+	{ LINUX_S_ISGID, S_ISGID },
+	{ LINUX_S_ISVTX, S_ISVTX },
 	{ 0, 0 }
 };
 
@@ -73,13 +76,6 @@ static void fix_perms(const char *cmd, const struct ext2_inode *inode,
 	struct utimbuf ut;
 	int i;
 
-	if (fd != -1)
-		i = fchmod(fd, mode_xlate(inode->i_mode));
-	else
-		i = chmod(name, mode_xlate(inode->i_mode));
-	if (i == -1)
-		com_err(cmd, errno, "while setting permissions of %s", name);
-
 #ifndef HAVE_FCHOWN
 	i = chown(name, inode_uid(*inode), inode_gid(*inode));
 #else
@@ -91,6 +87,13 @@ static void fix_perms(const char *cmd, const struct ext2_inode *inode,
 	if (i == -1)
 		com_err(cmd, errno, "while changing ownership of %s", name);
 
+	if (fd != -1)
+		i = fchmod(fd, mode_xlate(inode->i_mode));
+	else
+		i = chmod(name, mode_xlate(inode->i_mode));
+	if (i == -1)
+		com_err(cmd, errno, "while setting permissions of %s", name);
+
 	ut.actime = inode->i_atime;
 	ut.modtime = inode->i_mtime;
 	if (utime(name, &ut) == -1)
